feat: add OTLP/Traces output support

jjo opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines in the [CONTRIBUTING.md](https://github.com/falcosecurity/.github/blob/master/CONTRIBUTING.md) file and learn how to compile Falco from source [here](https://falco.org/docs/source).
2. Please label this pull request according to what type of issue you are addressing.
3. If the PR is unfinished while opening it specify a wip in the title before the actual title, for example, "wip: my awesome feature"
-->

**What type of PR is this?**
/kind feature

<!--
Please remove the leading whitespace before the `/kind <>` you uncommented.
-->

**Any specific area of the project related to this PR?**
/area outputs

<!--
Please remove the leading whitespace before the `/area <>` you uncommented.
-->

**What this PR does / why we need it**:
Implements OTLP/Traces (OpenTelemetry) output.

Folding Falco events into traces allows them to be represented as "hierarchical" views.
Using a TraceID taken from `container.id` (or any arbitrary output fields configured by the user), all Falco events will be "rooted" under the container (or fieldset) where it occurred, enabling a tree~ish view of these.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:
Will provide some screenshots  of these hierarchical views, as shown by an OTLP Traces backend like  [Jaeger](https://www.jaegertracing.io/) or [Grafana Tempo](https://grafana.com/oss/tempo/).

Issif commented:
Hi,

This is a terrific idea and PR!!! Thanks a lot. I'm in holidays til next Monday, I'll review asap.

jjo commented:
Using below `docker-compose.yaml`, and querying Grafana UI at <http://localhost:3000/explore> with Tempo datasource and `{}` query:
![image](https://github.com/falcosecurity/falcosidekick/assets/88727/8691eda5-9c38-489c-a280-98d9358fd4d4)

* On the left panel you see the four containers "containing" Falco events
* On the right, after clicking one of the above (chose the `event-generator`)

* `docker-compose.yaml`:
```
---
version: "3.9"
services:
  falco:
    image: falcosecurity/falco-no-driver:latest
    privileged: true
    command: "falco --modern-bpf -r /etc/falco/rules"
    volumes:
      - /var/run/docker.sock:/host/var/run/docker.sock
      - /dev:/host/dev
      - /proc:/host/proc:ro
      - /boot:/host/boot:ro
      - /lib/modules:/host/lib/modules:ro
      - ./etc/falco:/etc/falco:ro
  falcosidekick:
     #image: falcosecurity/falcosidekick:latest
     # NB: below image built from github.com/jjo/falcosidekick jjo/OTLP branch
    volumes:
      - ./etc/falco:/etc/falco:ro
    image: xjjo/falco-sidekick:2.28.0-jjo07
    command: -c /etc/falco/falcosidekick.yaml
    ports:
      - 2801:2801
    environment:
      # - OTEL_EXPORTER_OTLP_TRACES_ENDPOINT=http://jaeger:4318/v1/traces
      - OTEL_EXPORTER_OTLP_TRACES_ENDPOINT=http://tempo:4318/v1/traces
      - OTEL_EXPORTER_OTLP_TRACES_PROTOCOL=grpc
      - OTEL_EXPORTER_OTLP_HEADERS=X-Scope-OrgID=1
      - OTLP_TRACES_SYNCED=true
#  jaeger:
#    image: jaegertracing/all-in-one
#    command: [
#      "--collector.otlp.http.host-port=0.0.0.0:4318",
#      "--collector.otlp.http.tls.enabled=false"
#    ]
#    ports:
#      - 4318:4318
#      - 16686:16686
  tempo:
    image: grafana/tempo:latest
    ports:
      - 4317
      - 4318
      - 3200
    volumes:
      - ./etc/tempo:/etc/tempo:ro
    command: "-config.file /etc/tempo/config.yaml"
    restart: always
  grafana:
    image: grafana/grafana:10.0.3
    volumes:
      - ./etc/grafana/provisioning:/etc/grafana/provisioning:ro
    environment:
      - GF_AUTH_ANONYMOUS_ENABLED=true
      - GF_AUTH_ANONYMOUS_ORG_ROLE=Admin
      - GF_AUTH_DISABLE_LOGIN_FORM=true
      - GF_FEATURE_TOGGLES_ENABLE=traceqlEditor
    ports:
      - "3000:3000"
  event-generator:
    image: falcosecurity/event-generator
    command: run
    restart: always
  trigger:
    image: alpine
    command: ["sh", "-c", "while true; do cat /etc/shadow > /dev/null; sleep 5; done"]
```

jjo commented:
> Hi,
> 
> This is a terrific idea and PR!!! Thanks a lot. I'm in holidays til next Monday, I'll review asap.

Huh 1st thought it was kinda crazy to fold these as Traces, much appreciate your feedback!

jjo commented:
And a Jaeger screenshot (using same `docker-compose.yaml` file, edited to enable Jaeger and disable Tempo)
1) UI landing:
![image](https://github.com/falcosecurity/falcosidekick/assets/88727/c0014030-8285-4b46-a500-977b80c6cfe9)
 
2) After clicking one the trace corresponding to the `event-generator`
![image](https://github.com/falcosecurity/falcosidekick/assets/88727/6ccab252-6b6e-4463-9321-7d043dfc0127)

Issif commented on main.go:
```diff
@@ -730,12 +731,29 @@ func init() {
+		if err != nil {
+			config.OTLP.Traces.Endpoint = ""
+		} else {
+			outputs.EnabledOutputs = append(outputs.EnabledOutputs, "OTLP.Traces")
```
```suggestion
			outputs.EnabledOutputs = append(outputs.EnabledOutputs, "OTLPTraces")
```

Issif commented on main.go:
```diff
@@ -730,12 +731,29 @@ func init() {
+			config.OTLP.Traces.Endpoint = ""
+		} else {
+			outputs.EnabledOutputs = append(outputs.EnabledOutputs, "OTLP.Traces")
+			log.Printf("[INFO] : OTLP.Traces=%+v\n", config.OTLP.Traces)
```
```suggestion
```

Issif commented on handlers.go:
```diff
@@ -402,4 +402,8 @@ func forwardEvent(falcopayload types.FalcoPayload) {
 	}
+
+	if config.OTLP.Traces.Endpoint != "" && (falcopayload.Priority >= types.Priority(config.OTLP.Traces.MinimumPriority)) {
+		go otlpClient.OTLPPost(falcopayload)
```
I guess we may have more than traces in the future, so the function should be `OTLPTracesPost()`, wdyt?

jjo replied:
+1

Issif commented on otlp_init.go:
```diff
@@ -0,0 +1,125 @@
+
+// See https://opentelemetry.io/docs/concepts/sdk-configuration/otlp-exporter-configuration/
+// FYI for traces, you can also use:
+// - OTEL_EXPORTER_OTLP_HEADERS, OTEL_EXPORTER_OTLP_TRACES_HEADERS
+// - OTEL_EXPORTER_OTLP_TIMEOUT, OTEL_EXPORTER_OTLP_TRACES_TIMEOUT
+// - OTEL_EXPORTER_OTLP_PROTOCOL, OTEL_EXPORTER_OTLP_TRACES_PROTOCOL
```
do you you think we should expose the settings too?

Issif commented on config.go:
```diff
@@ -32,8 +32,10 @@ func getConfig() *types.Configuration {
+		OTLP:            types.OTLPOutputConfig{},
 	}
 
+	otlpSetEnvs()
```
why did you choose to create a dedicated function and not doing like for the others ouputs?

jjo replied:
mainly because of the complexity added from `OTEL_*` native SDK env vars vs `OTLP_*` introduced here (+ its auto-config machinery) and the needed ~mapping between them.

Issif replied:
I see, is there a way to explicitly list all of them and expose them in the config file?

jjo replied:
Unsure tbh, although I see your point, e.g. ppl wanting to set headers via config rather than being "forced" to set OTEL_EXPORTER_OTLP_TRACES_HEADERS.

Recall I did try to craft above headers under `viper` config machinery, but ended up needing to create a golang type hierarchy at each `_` mapped to e.g. `Otel.Exporter.Otlp.Traces.Headers` equivalent types "nesting":

1) Do you know a way around this (a nested type at each `_` from the env)?
2) Alternatively, we could "shortcut" these the same way as I did with `OTLP_TRACES_ENDPOINT` (own, yaml config supported) <-> `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT` or `OTEL_EXPORTER_OTLP_ENDPOINT` (SDK) as e.g.:
   * `OTLP_TRACES_HEADERS` (own) <-> `OTEL_EXPORTER_OTLP_TRACES_HEADERS, OTEL_EXPORTER_OTLP_HEADERS`
   * `OTLP_TRACES_TIMEOUT` (own) <-> _[ditto]_
   * `OTLP_TRACES_PROTOCOL` (own) <-> _[ditto]_

wdyt?

Issif commented on otlp_init.go:
```diff
@@ -0,0 +1,125 @@
+		if otlpOS.Getenv(v.EnvName) != "" {
+			value = otlpOS.Getenv(v.EnvName) + v.Path
+			otlpOS.Setenv(v.Target, value)
+			break
```
why `break`? it breaks the loop, it should be a `continue` to pass to next env var

jjo replied:
^- 861ceee9

Issif commented:
I did a first review, still have to read the whole logic and understand how it works

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,157 @@
+func (c *Client) newTrace(falcopayload types.FalcoPayload) *trace.Span {
+	traceID, _, err := generateTraceID(falcopayload, c.Config)
+	if err != nil {
+		log.Printf("[ERROR] : Error generating trace id: %v for output fields %v", err, falcopayload.OutputFields)
```
```suggestion
		log.Printf("[ERROR] : OLTP Traces : Error generating trace id: %v for output fields %v", err, falcopayload.OutputFields)
```

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,157 @@
+	span.End(trace.WithTimestamp(endTime))
+
+	if c.Config.Debug {
+		log.Printf("[DEBUG] : OTLP payload generated successfully for traceid=%s", span.SpanContext().TraceID())
```
```suggestion
		log.Printf("[DEBUG] : OTLP Traces - payload generated successfully for traceid=%s", span.SpanContext().TraceID())
```

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,157 @@
+// OTLPPost generates an OTLP trace _implicitly_ via newTrace() by
+// calling OTEL SDK's tracer.Start() --> span.End(), i.e. no need to explicitly
+// do a HTTP POST
+func (c *Client) OTLPPost(falcopayload types.FalcoPayload) {
```
```suggestion
func (c *Client) OTLPTracesPost(falcopayload types.FalcoPayload) {
```

because other OTLP implementations may be created in future

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,157 @@
+		go c.CountMetric(Outputs, 1, []string{"output:otlp/traces", "status:error"})
+		c.Stats.OTLP.Add(Error, 1)
+		c.PromStats.Outputs.With(map[string]string{"destination": "otlp/traces", "status": Error}).Inc()
+		log.Printf("[ERROR] : OTLPError generating trace")
```
```suggestion
		log.Printf("[ERROR] : OLTP Traces - Error generating trace")
```

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,157 @@
+func sanitizeOutputFields(falcopayload types.FalcoPayload) map[string]interface{} {
+	ret := make(map[string]interface{})
+	for k, v := range falcopayload.OutputFields {
+		k := strings.Replace(k, ".", "_", -1)
```
```suggestion
	k := strings.ReplaceAll(k, ".", "_")
```

Issif commented on otlp_init.go:
I think this file should be in the `outputs` folder, and the management of the ctx, close, etc, declared with the client creation.

jjo replied:
+1 moving it there in next commit.

jjo replied:
+1 will move

jjo replied:
^- [3ac1ad4](https://github.com/falcosecurity/falcosidekick/pull/613/commits/3ac1ad465ef0b5cb8c4860b455d18c4167193b51)

poiana commented:
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falcosidekick/pull/613#" title="Author self-approved">jjo</a>*
**Once this PR has been reviewed and has the lgtm label**, please ask for approval from issif. For more information see [the Kubernetes Code Review Process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process).

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalcosidekick).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/falcosecurity/falcosidekick/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":["issif"]} -->

jjo commented on Dockerfile:
```diff
@@ -3,7 +3,7 @@ ARG BASE_IMAGE=alpine:3.17
 LABEL MAINTAINER "Thomas Labarussias <issif+falcosidekick@gadz.org>"
 
-RUN apk add --update --no-cache ca-certificates
+RUN apk add --update --no-cache ca-certificates gcompat
```
The resulting binary from compiling it locally:
```
$ go version  
go version go1.20.5 linux/amd64
$ dpkg -l libc6
ii  libc6:amd64    2.35-0ubuntu3.1 amd64        GNU C Library: Shared libraries

# Locally ok
$ ldd falcosidekick
        linux-vdso.so.1 (0x00007ffe51de6000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f0c31200000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f0c31449000)

# Missing symbols from stock alpine
$ docker run -v $PWD:/app alpine ldd  /app/falcosidekick
        /lib64/ld-linux-x86-64.so.2 (0x7fbe539a0000)
        libc.so.6 => /lib64/ld-linux-x86-64.so.2 (0x7fbe539a0000)
Error relocating /app/falcosidekick: __vfprintf_chk: symbol not found
Error relocating /app/falcosidekick: __fprintf_chk: symbol not found
```
, then answers like [this one](https://stackoverflow.com/a/70256941/9443059) after searching from such errors.

Issif replied:
I see, we just need to be sure to not break any compatibility with arm systems.

Issif commented on main.go:
```diff
@@ -730,12 +731,29 @@ func init() {
+		} else {
+			outputs.EnabledOutputs = append(outputs.EnabledOutputs, "OTLP.Traces")
+			log.Printf("[INFO] : OTLP.Traces=%+v\n", config.OTLP.Traces)
+			otlpShutdown = otlpInit()
```
should be done with the creation of the client

jjo replied:
yeah much better, will re-factor it together with `otlp_init.go` move to `outputs/`

jjo replied:
^- [3ac1ad4](https://github.com/falcosecurity/falcosidekick/pull/613/commits/3ac1ad465ef0b5cb8c4860b455d18c4167193b51)

Issif commented on main.go:
```diff
@@ -730,12 +731,29 @@ func init() {
 
+	if config.OTLP.Traces.Endpoint != "" {
+		var err error
+		otlpClient, err = outputs.NewClient("OTLP", config.OTLP.Traces.Endpoint, false, false, config, stats, promStats, statsdClient, dogstatsdClient)
```
After deeper review, This creation of a client is useless, the `NewClient` is only for output that manages the payload as a json. You create a specific client for OTLP, you should see how we did it for AWS

jjo replied:
^- 53edd969

jjo commented on config.go:
```diff
@@ -475,6 +477,15 @@ func getConfig() *types.Configuration {
+	v.SetDefault("OTLP.Traces.Synced", false)
+	v.SetDefault("OTLP.Traces.MinimumPriority", "")
+	v.SetDefault("OTLP.Traces.Insecure", false)
+	v.SetDefault("OTLP.Traces.TraceIDHash", "")
+	// NB: Unfortunately falco events don't provide endtime, artificially set
+	// it to 1000ms by default, override-able via OTLP_DURATION environment variable.
+	v.SetDefault("OTLP.Traces.Duration", 1000)
```
+1 didn't realize that section, adding in next commit

jjo commented on README.md:
```diff
@@ -1215,7 +1219,11 @@ order is
 - **DYNATRACE_CHECKCERT** : check if ssl certificate of the output is valid (default: `true`)
 - **DYNATRACE_MINIMUMPRIORITY** : minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
-
+- **OTLP_TRACES_ENDPOINT**: OTLP endpoint, if not empty, OpenTelemetry traces output is enabled
```
done at 53edd969

Issif commented on handlers.go:
```diff
@@ -402,4 +402,8 @@ func forwardEvent(falcopayload types.FalcoPayload) {
 		go dynatraceClient.DynatracePost(falcopayload)
 	}
+
+	if config.OTLP.Traces.Endpoint != "" && (falcopayload.Priority >= types.Priority(config.OTLP.Traces.MinimumPriority)) {
```
```suggestion
	if config.OTLP.Traces.Endpoint != "" && (falcopayload.Priority >= types.Priority(config.OTLP.Traces.MinimumPriority))  || falcopayload.Rule == testRule) {
```

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+var getTracerProvider = otel.GetTracerProvider
+
+func NewOtlpTracesClient(config *types.Configuration, stats *types.Statistics, promStats *types.PromStatistics, statsdClient, dogstatsdClient *statsd.Client) (*Client, error) {
+	otlpClient, err := NewClient("OTLP.Traces", config.OTLP.Traces.Endpoint, false, false, config, stats, promStats, statsdClient, dogstatsdClient)
```
this creates an http client, in the settings I also see gRPC. I think we can keep only http for start, wdyt?

Issif replied:
```suggestion
	otlpClient, err := NewClient("OTLP.Traces", config.OTLP.Traces.Endpoint, false, config.OTLP.Traces.CheckCert, config, stats, promStats, statsdClient, dogstatsdClient)
```

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+	if err != nil {
+		return nil, err
+	}
+	log.Printf("[INFO] : OTLP.Traces=%+v\n", config.OTLP.Traces)
```
```suggestion
	log.Printf("[INFO]  : OLTP Traces - %+v\n", config.OTLP.Traces)
```

Issif commented on types/types.go:
```diff
@@ -800,6 +801,7 @@ type Statistics struct {
 	N8N               *expvar.Map
 	OpenObserve       *expvar.Map
 	Dynatrace         *expvar.Map
+	OTLP              *expvar.Map
```
```suggestion
	OTLPTraces        *expvar.Map
```

Issif commented on stats.go:
```diff
@@ -81,6 +81,7 @@ func getInitStats() *types.Statistics {
 		N8N:               getOutputNewMap("n8n"),
 		OpenObserve:       getOutputNewMap("openobserve"),
 		Dynatrace:         getOutputNewMap("dynatrace"),
+		OTLP:              getOutputNewMap("otlp"),
```
```suggestion
		OTLPTraces:        getOutputNewMap("otlp"),
```

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+	if trace == nil {
+		go c.CountMetric(Outputs, 1, []string{"output:otlp/traces", "status:error"})
+		c.Stats.OTLP.Add(Error, 1)
+		c.PromStats.Outputs.With(map[string]string{"destination": "otlp/traces", "status": Error}).Inc()
```
```suggestion
		c.PromStats.Outputs.With(map[string]string{"destination": "otlptraces", "status": Error}).Inc()
```

not sure the / is accepted, anyway,  to keep consistency with other outputs

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+		return
+	}
+	// Setting the success status
+	go c.CountMetric(Outputs, 1, []string{"output:otlp/traces", "status:ok"})
```
```suggestion
	go c.CountMetric(Outputs, 1, []string{"output:otlptraces", "status:ok"})
```

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+
+	trace := c.newTrace(falcopayload)
+	if trace == nil {
+		go c.CountMetric(Outputs, 1, []string{"output:otlp/traces", "status:error"})
```
```suggestion
		go c.CountMetric(Outputs, 1, []string{"output:otlptraces", "status:error"})
```

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+	// Setting the success status
+	go c.CountMetric(Outputs, 1, []string{"output:otlp/traces", "status:ok"})
+	c.Stats.OTLP.Add(OK, 1)
+	c.PromStats.Outputs.With(map[string]string{"destination": "otlp/traces", "status": OK}).Inc()
```
```suggestion
	c.PromStats.Outputs.With(map[string]string{"destination": "otlptraces", "status": OK}).Inc()
```

Issif commented on config.go:
```diff
@@ -740,6 +803,11 @@ func getConfig() *types.Configuration {
 	c.Googlechat.MessageFormatTemplate = getMessageFormatTemplate("Googlechat", c.Googlechat.MessageFormat)
 	c.Cliq.MessageFormatTemplate = getMessageFormatTemplate("Cliq", c.Cliq.MessageFormat)
+	c.OTLP.Traces.Endpoint = strings.TrimSpace(c.OTLP.Traces.Endpoint)
+	if c.OTLP.Traces.TraceIDHash != "" {
+		c.OTLP.Traces.TraceIDHashTemplate = getMessageFormatTemplate("OTLP", c.OTLP.Traces.TraceIDHash).Option("missingkey=zero")
+	}
```
Is it really a requested feature? Can't we just enforce it by our own? It's not clear to me which is the benefit

jjo replied:
I found it to be safer to let the gotemplate engine fallback to `""` instead of the verbatim `"<no value>"` if that option is not set, see https://pkg.go.dev/text/template#Template.Option.

Issif commented on config.go:
```diff
@@ -476,6 +515,30 @@ func getConfig() *types.Configuration {
 	v.SetDefault("Dynatrace.CheckCert", true)
 	v.SetDefault("Dynatrace.MinimumPriority", "")
 
+	v.SetDefault("OTLP.Traces.Endpoint", altEnvs([]string{
+		"OTEL_EXPORTER_OTLP_TRACES_ENDPOINT",
+		"OTEL_EXPORTER_OTLP_ENDPOINT+/v1/traces",
+	}))
+	v.SetDefault("OTLP.Traces.Protocol", altEnvs([]string{
+		"OTEL_EXPORTER_OTLP_TRACES_PROTOCOL",
+		"OTEL_EXPORTER_OTLP_PROTOCOL",
+	}))
+	v.SetDefault("OTLP.Traces.Headers", altEnvs([]string{
+		"OTEL_EXPORTER_OTLP_TRACES_HEADERS",
+		"OTEL_EXPORTER_OTLP_HEADERS",
+	}))
+	v.SetDefault("OTLP.Traces.Timeout", altEnvs([]string{
+		"OTEL_EXPORTER_OTLP_TRACES_TIMEOUT",
+		"OTEL_EXPORTER_OTLP_TIMEOUT",
+	}))
```
I understand these envs vars are the default ones used by the sdk, but I prefer to expose them in the settings as specific settings. It seems to repeat things but it will clearer for users and also easier for the implementation in the helm chart later.

Moreover, for `OTLP.Traces.Timeout` it can't work, because it's an `int64` and `altEnvs()` returns string.

I understand what you're trying to do, but it will introduce too much complexity when setting all explicitly could work.

Issif replied:
```suggestion
	v.SetDefault("OTLP.Traces.Endpoint", "")
	v.SetDefault("OTLP.Traces.Protocol", "http")
	v.SetDefault("OTLP.Traces.Headers", "")
	v.SetDefault("OTLP.Traces.Timeout", 1000)
```

jjo replied:
> why are you using a slice of string and not a map? i think the iteration over a slice is not consistent, the order is not always kept, it is in a map

(?) I think is the other way round, as a slice is essentially a header that specifies a range for an underlying array, from https://go.dev/ref/spec#For_statements ->

1. For an array, pointer to array, or slice value a, the **index iteration values are produced in increasing order**, starting at element index 0. If at most one iteration variable is present, the range loop produces iteration values from 0 up to len(a)-1 and does not index into the array or slice itself. For a nil slice, the number of iterations is 0.
2. [...]
3. The **iteration order over maps is not specified and is not guaranteed to be the same from one iteration to the next**. If a map entry that has not yet been reached is removed during iteration, the corresponding iteration value will not be produced. If a map entry is created during iteration, that entry may be produced during the iteration or may be skipped. The choice may vary for each entry created and from one iteration to the next. If the map is nil, the number of iterations is 0.

jjo replied:
> I understand these envs vars are the default ones used by the sdk, but I prefer to expose them in the settings as specific settings. It seems to repeat things but it will clearer for users and also easier for the implementation in the helm chart later.
> 

I'd rather advocate for this one: I do want to keep `OTEL_*` envvars compatibility, to let people already working with OTEL and familiar with those env vars, to let the sidekick behave as any other OTEL application (mind that the OTEL envvars spec is cross-implementation also) -- can think of known OTEL env vars already being set from Kubernetes secrets for example.

> Moreover, for `OTLP.Traces.Timeout` it can't work, because it's an `int64` and `altEnvs()` returns string.

+1 can fix it of course (if you agree to keeping this OTEL env vars handling)

> 
> I understand what you're trying to do, but it will introduce too much complexity when setting all explicitly could work.

🤷🏼 I know, but IMO the tradeoff to let this behave as any other OTEL app weights more =)

jjo replied:
> Moreover, for OTLP.Traces.Timeout it can't work, because it's an int64 and altEnvs() returns string.

BTW this does work, note the  `OTEL_EXPORTER_OTLP_TRACES_TIMEOUT=4242` env being set ->
```
$ go build -o falcosidekick
$ OTLP_TRACES_SYNCED=true \
  OTLP_TRACES_CHECKCERT=false \
  OTLP_TRACES_ENDPOINT=http://172.17.0.1:4318/v1/traces \
  OTEL_EXPORTER_OTLP_TRACES_TIMEOUT=4242 ./falcosidekick
2023/09/06 18:59:01 [INFO] : OTLP.Traces={Endpoint:http://172.17.0.1:4318/v1/traces Protocol: Timeout:4242 Headers: Duration:1000 CheckCert:false Synced:true MinimumPriority: TraceIDHash: TraceIDHashTemplate:<nil>}
2023/09/06 18:59:01 [INFO]  : Falco Sidekick version: devel
2023/09/06 18:59:01 [INFO]  : Enabled Outputs : [OTLPTraces]
2023/09/06 18:59:01 [INFO]  : Falco Sidekick is up and listening on :2801
```

Issif commented on outputs/otlp_init.go:
```diff
@@ -0,0 +1,84 @@
+)
+
+const (
+	OTLPinstrumentationName    = "falcosidekick.otlp"
```
I prefer to set `falco`, as it's the source.

jjo replied:
+1, done at a7865a5e, s/shot (using the above `docker-compose.yaml`) just for completeness
![image](https://github.com/falcosecurity/falcosidekick/assets/88727/b16cff5d-44bc-4522-9a2b-3bee48559bb5)

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+	}
+	// Setting the success status
+	go c.CountMetric(Outputs, 1, []string{"output:otlp/traces", "status:ok"})
+	c.Stats.OTLP.Add(OK, 1)
```
```suggestion
	c.Stats.OTLPTraces.Add(OK, 1)
```

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+	trace := c.newTrace(falcopayload)
+	if trace == nil {
+		go c.CountMetric(Outputs, 1, []string{"output:otlp/traces", "status:error"})
+		c.Stats.OTLP.Add(Error, 1)
```
```suggestion
		c.Stats.OTLPTraces.Add(Error, 1)
```

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+// calling OTEL SDK's tracer.Start() --> span.End(), i.e. no need to explicitly
+// do a HTTP POST
+func (c *Client) OTLPTracesPost(falcopayload types.FalcoPayload) {
+	c.Stats.OTLP.Add(Total, 1)
```
```suggestion
	c.Stats.OTLPTraces.Add(Total, 1)
```

jjo replied:
(already committed)

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+		log.Printf("[ERROR] : OLTP Traces - Error generating trace")
+		return
+	}
+	// Setting the success status
```
I don't see where the trace is sent

jjo replied:
indeed, it's the "magic" from the underlying OTEL libraries: by just "ending" the Span in the trace -- it'll get buffered (by default, controlled by  `config.OTLP.Traces.Synced` defaulting to `false`), then later `Post`-ed.

Issif replied:
It sounds then more logic to me to return the `*trace.Span` and Ending it in the `OTLPTracesPost()` method. It doesn't seem to return an error, didn't find how to change that in the SDK. How are we sure the trace has been correctly updated with the span?

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+
+const (
+	templateOption       = "missingkey=zero"
+	kubeTemplateStr      = `{{.cluster}}{{.k8s_pod_name}}{{.k8s_ns_name}}{{.k8s_container_name}}`
+	containerTemplateStr = `{{.container_id}}`
```
where do you get the `.cluster` from? btw, you only address syscalls events, in a pod/container, but we need to address other use cases, like the events from the plugins, without the fields you use.

jjo replied:
well ... under the umbrella of «opinionated» :P ... srsly speaking, using a rather _common_ `cluster` label, but I see your point, I'm fine removing it and keeping `{{container.id}}` only as default (still letting ppl override it via `config.OTLP.Traces.TradeIDHash`).

Issif replied:
even if people are able to override the template, we still have the issue that the available fields are not the same between the sources

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+	return traceID, tplStr, err
+}
+
+func randomHex(n int) (string, error) {
+	bytes := make([]byte, n)
+	if _, err := rand.Read(bytes); err != nil {
+		return "", err
+	}
+	return hex.EncodeToString(bytes), nil
+}
```
To generate an UUID for each event, I'm using [github.com/google/uuid](https://github.com/google/uuid). See https://github.com/falcosecurity/falcosidekick/blob/5cfd355809f38d990b5458f354b091a653af7798/handlers.go#L102

It could replace this whole function and have a simpler code, wdyt?

jjo replied:
Well nopes directly, I tried and failed, as the traceID needs to be _exactly_ 32chars hexa.
That said, we could re-use that UUID and massage it (remove the `-`, and cut it from 36chars to 32chars).
Opinion?

jjo replied:
hm actually, it was indeed pretty easy to just use UUID string as "seed" for the traceID in this case (no value from template rendering), implemented at a45a91da

jjo commented on outputs/otlp.go:
```diff
@@ -0,0 +1,172 @@
+	return otlpClient, nil
+}
+
+// newTrace returns a new Trace object.
```
tricky one: it's a Trace (that tries to re-use a "consistent" traceID across further spans from e.g. same `container.id`) containing a single Span.

poiana commented:
Adding label `do-not-merge/contains-merge-commits` because PR contains merge commits, which are not allowed in this repository.
Use `git rebase` to reapply your commits on top of the target branch. Detailed instructions for doing so can be found [here](https://git.k8s.io/community/contributors/guide/github-workflow.md#4-keep-your-branch-in-sync).

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

jjo commented:
> Adding label `do-not-merge/contains-merge-commits` because PR contains merge commits, which are not allowed in this repository. Use `git rebase` to reapply your commits on top of the target branch. Detailed instructions for doing so can be found [here](https://git.k8s.io/community/contributors/guide/github-workflow.md#4-keep-your-branch-in-sync).

Thanks!, just done it and repushed ✅

jjo commented:
@Issif 
> What I mean is, do we really need to allow the user to specify the value they want to use? Like I said in another comment, >container.id or k8s.pod.name are not always present in the payload (if the source is a plugin).

IMO it's a _key_ feature actually, indeed for those cases were people may want to use their own label(set) as traceID seed, to let traces be "rooted" against it as 1:1.

jjo commented:
> @Issif
> 
> > What I mean is, do we really need to allow the user to specify the value they want to use? Like I said in another comment, >container.id or k8s.pod.name are not always present in the payload (if the source is a plugin).
> 
> IMO it's a _key_ feature actually, indeed for those cases were people may want to use their own label(set) as traceID seed, to let traces be "rooted" against it as 1:1.

That said, +1 on removing the Kubernetes bits as they're are "too much" opinionated, i.e. assumes some external `cluster` tag addition (which is a common case fwiw for multi-cluster setups, anyway still modifiable by config).

jjo commented:
@Issif
> It sounds then more logic to me to return the *trace.Span and Ending it in the OTLPTracesPost() method. It doesn't seem to return an error, didn't find how to change that in the SDK. How are we sure the trace has been correctly updated with the span?

Unsure if it's worth tbh, as the only caller is `handlers.go`, where every output handler is called as `go FooPost()` (i.e. return value ignored).

jjo commented:
@Issif PTAL, replied to your latest feedback, changes since your last review:
* removed `{{.cluster}}{{.k8s_pod_name}}{{.k8s_ns_name}}{{.k8s_container_name}}` opinionated/special-case for `OTLP_TRACES_TRACEIDHASH`, leaving only `{{container_id}}` as default but only used if it _actually_ produces a rendered non empty output (i.e. `{{container.id}}` present), else falls back to random traceID.
* rebased (fixed go.mod conflicts)

Issif commented on README.md:
```diff
@@ -666,6 +670,18 @@ dynatrace:
+otlp:
+  traces:
+    # endpoint: "" # OTLP endpoint in the form of http://{domain or ip}:4318/v1/traces
+    # protocol: "" # OTLP protocol: http or grpc
+    # timeout: "" # OTLP timeout: timeout value in milliseconds.
+    # headers: "" # OTLP headers: list of headers to apply to all outgoing traces in the form of "some-key=some-value,other-key=other-value"
+    # synced: false # Set to true if you want traces to be sent synchronously
+    # minimumpriority: "" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+    # checkcert: false # Set if you want to skip TLS certificate validation
```
Can you specify the default values please

jjo replied:
Done, pushing in next commit.

Issif commented on README.md:
```diff
@@ -1218,7 +1234,15 @@ order is
+- **OTLP_TRACES_HEADERS**: OTLP headers: list of headers to apply to all outgoing traces in the form of "some-key=some-value,other-key=other-value"
+- **OTLP_TRACES_MINIMUMPRIORITY**: minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+- **OTLP_TRACES_SYNCED**: synced OTLP traces, instead of batched (default: `false`)
+- **OTLP_TRACES_CHECKCERT**: check if ssl certificate of the endpoint is valid (default: `false`)
```
by default, the `checkcert` is true

jjo replied:
Fixed, pushing in next commit.

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,150 @@
+	}
+	buf := &bytes.Buffer{}
+	if err := tpl.Execute(buf, outputFields); err != nil {
+		log.Printf("[WARNING] : OTLP - Error expanding template: %v", err)
```
```suggestion
		log.Printf("[WARNING] : OTLP Traces - Error expanding template: %v", err)
```

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,150 @@
+	var err error
+	traceIDStr, tplStr := renderTraceIDFromTemplate(falcopayload, config)
+
+	switch traceIDStr {
+	case "":
+	case "<no value>":
+		// Template produced no string, derive the traceID from the payload UUID
+		traceIDStr = falcopayload.UUID
+	}
```
```suggestion
	if traceIDStr == "<no value>" {
		// Template produced no string, derive the traceID from the payload UUID
		traceIDStr = falcopayload.UUID
	}
```

Issif commented on outputs/otlp.go:
```diff
@@ -0,0 +1,150 @@
+	tplStr := config.OTLP.Traces.TraceIDHash
+	tpl := config.OTLP.Traces.TraceIDHashTemplate
+	outputFields := sanitizeOutputFields(falcopayload)
+	// Default to container.id `{{.container_id}}` as templating "seed" to generate traceID.
+	if tplStr == "" {
+		tpl, tplStr = containerTemplate, containerTemplateStr
+	}
```
Once again, we can't rely on that field as some rules don't output it or it's not available for plugins.

Can we allow the users to specify multiple seeds? And use them in an order.

k8s.pod.name + container.id > k8s.pod.name > container.id > container.name > ???

Still remains the case with events without these fields. We can also choose to explicitly consider only the syscalls events.

jjo replied:
Understood and acknowledged already, my take is that by exposing it as a env var, we'd let users to adapt it to their own setup (including plugins).
Still, +1 on changing the default to something that would have a "wider coverage", IMO we could put all the above into the single existing default, as we're here chasing a seed that we'll hash into the `traceID`, to set as _the_ rooting value (for the spans).

What do you think about just changing the default? ->
* from: `{{default_id}}`
* to: `{{k8s_ns_name}}{{k8s_pod_name}}{{container_name}}{{container_id}}`

NOTE that we can't _easily_ use dots in `gotemplate`, as it'd assume fields hierarchy and make the template _really_ cumbersome (need to use gotemplate nested `index` calls, etc).

Alternatively, I could implement your suggestion as a multi-valued string e.g. separated by `;` (single envvar / config field), but IMO is not needed as per the "aggregated" nature of the template rendering.

jjo replied:
Actually went ahead an implemented the above "expanded" default template string in fe6b950c, for your consideration.

Issif commented on docs/examples/otlp/README.md:
```diff
@@ -0,0 +1,44 @@
+## What
```
Even if I understand why you did that, can you remove the whole example folder please. I prefer we all think about what we want to do before integrating so much new things in the repo. Especially it's not a good idea to embed falco rules there.

jjo replied:
Done at 289f2f06.

Issif commented:
@jjo I will work this you on your PR, thanks for your patience

jjo commented:
> @jjo I will work this you on your PR, thanks for your patience

thanks

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [d258885](https://github.com/falcosecurity/falcosidekick/commits/d258885e820d2964ad37ac2df7fb4a5534e2bb7c) Merge pull request #2 from Issif/jjo/OTLP

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

poiana commented:
Thanks for your pull request. Before we can look at it, you'll need to add a 'DCO signoff' to your commits.

:memo: **Please follow instructions in the [contributing guide](https://github.com/falcosecurity/.github/blob/main/CONTRIBUTING.md) to update your commits with the DCO**

Full details of the Developer Certificate of Origin can be found at [developercertificate.org](https://developercertificate.org/).

**The list of commits missing DCO signoff**:

- [e37e89b](https://github.com/falcosecurity/falcosidekick/commits/e37e89b6237721e33ebf3a3b895bddea86649745) merge issif's https://github.com/jjo/falcosidekick/pull/2 + rebase

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

Issif commented on docs/outputs/otlp_traces.md:
````diff
@@ -0,0 +1,54 @@
+    # minimumpriority: "" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+    # checkcert: true # Set if you want to skip TLS certificate validation (default: true)
+```
+
````
```suggestion

## Additional info

> **Note**
The OTLP Traces are only available for the source: `syscalls`.

```

Issif reviewed (changes requested):
Did you test my changes? I guess we're closed to the merge, can you rebase on the master and squash your commits after having applied my suggestions. thanks

jjo commented on docs/outputs/otlp.md:
```diff
@@ -0,0 +1,63 @@
+## Example docker-compose
+
+At [../examples/otlp/](../examples/otlp/) you'll find an example
+`docker-compose.yaml` setup to run an example OTLP traces setup.
```
+1 will do

jjo replied:
added at 7ddf7a7b

jjo commented on outputs/otlp.go:
```diff
@@ -0,0 +1,150 @@
+	tplStr := config.OTLP.Traces.TraceIDHash
+	tpl := config.OTLP.Traces.TraceIDHashTemplate
+	outputFields := sanitizeOutputFields(falcopayload)
+	// Default to container.id `{{.container_id}}` as templating "seed" to generate traceID.
```
resolving as already fixed by your recent PR

jjo commented on config_example.yaml:
```diff
@@ -483,3 +483,14 @@ dynatrace:
+    # minimumpriority: "" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+    # checkcert: true # Set if you want to skip TLS certificate validation (default: true)
+    # duration: 1000 # Artificial span duration in milliseconds (default: 1000)
+    # traceidhash: "" # Trace ID Hash "seed" as gotemplate, to select fields from falcopayload output fields (default: `{{container_id}}` to use `container.id`)
```
resolving as N/A'd by your recent PR

jjo commented:
> Did you test my changes? I guess we're closed to the merge, can you rebase on the master and squash your commits after having applied my suggestions. thanks

I'll have sometime today to test the whole stack, will let you know how it goes.
Regarding the (commented-out) tests:
* commented-out ones where mostly to handle templating, which is now N/A
* to avoid having to "record" actual hash values, I had taken the approach to test for inter- test-cases diff, but I see that you've chosen the former (fine w/that)
=> don't think there's further work worth doing on the unit test side.

Issif reviewed (commented):
can you fix the CI errors and squash your commits please

jjo commented:
> can you fix the CI errors and squash your commits please

on it

Issif commented on docs/outputs/otlp_traces.md:
````diff
@@ -0,0 +1,246 @@
+      # OTEL_EXPORTER_OTLP_TIMEOUT: 10000
+    # minimumpriority: "" # minimum priority of event for using this output, order is emergency|alert|critical|error|warning|notice|informational|debug or "" (default)
+    # checkcert: true # Set if you want to skip TLS certificate validation (default: true)
+```
````
Can you add some screenshots in different tools? like you did in the PR description. See the other output docs for ref. 
Thanks

jjo replied:
+1 added at 9353d05

jjo commented on config.go:
```diff
@@ -36,6 +36,12 @@ import (
 	"github.com/falcosecurity/falcosidekick/types"
 )
 
+// NB: create OS interface to allow unit-testing
+type OS interface {
+	Getenv(string) string
+	Setenv(string, string) error
+}
```
correct, removing.

jjo commented on types/types.go:
```diff
@@ -763,6 +764,26 @@ type OpenObserveConfig struct {
+	ExtraEnvVars        map[string]string
+	CheckCert           bool
+	MinimumPriority     string
+	TraceIDHash         string
+	TraceIDHashTemplate *template.Template
```
+1 removing

jjo commented on config.go:
```diff
@@ -501,6 +508,18 @@ func getConfig() *types.Configuration {
+	v.SetDefault("OTLP.Traces.Synced", false)
+	v.SetDefault("OTLP.Traces.MinimumPriority", "")
+	v.SetDefault("OTLP.Traces.CheckCert", true)
+	v.SetDefault("OTLP.Traces.TraceIDHash", "")
```
+1 removing

jjo commented on outputs/otlp_test.go:
```diff
@@ -0,0 +1,183 @@
+// func (*MockTracerProvider) Tracer(string, ...trace.TracerOption) trace.Tracer {
+// 	return &MockTracer{}
+// }
+
+// // Tracer interface
+// func (*MockTracer) Start(ctx context.Context, name string, opts ...trace.SpanStartOption) (context.Context, trace.Span) {
+// 	return ctx, &MockSpan{
+// 		ctx:        ctx,
+// 		name:       name,
+// 		startOpts:  opts,
+// 		attributes: make(map[attribute.Key]attribute.Value),
+// 	}
+// }
+
+// // Span interface
+// func (*MockSpan) AddEvent(string, ...trace.EventOption)            {}
+// func (*MockSpan) IsRecording() bool                                { return true }
+// func (*MockSpan) RecordError(err error, opts ...trace.EventOption) {}
+// func (*MockSpan) SetName(name string)                              {}
+// func (*MockSpan) SetStatus(code codes.Code, description string)    {}
+
+// func (*MockSpan) TracerProvider() trace.TracerProvider { return &MockTracerProvider{} }
+
+// func (m *MockSpan) End(opts ...trace.SpanEndOption) {
+// 	m.endOpts = opts
+// }
+
+// func (m *MockSpan) SetAttributes(kv ...attribute.KeyValue) {
+// 	for _, k := range kv {
+// 		m.attributes[k.Key] = k.Value
+// 	}
+// }
+
+// func (m *MockSpan) SpanContext() trace.SpanContext {
+// 	return trace.SpanContextFromContext(m.ctx)
+// }
+
+// func MockGetTracerProvider() trace.TracerProvider {
+// 	return &MockTracerProvider{}
+// }
```
+1 doing

jjo commented:
> > Did you test my changes? I guess we're closed to the merge, can you rebase on the master and squash your commits after having applied my suggestions. thanks
> 
> @Issif my tests were not working because of your added `"syscalls"` filter in this line https://github.com/falcosecurity/falcosidekick/pull/613/files#diff-56307bdb7ded67ebe516f562b90384407b21b2d34e8b5766b42602e1fb9705f0R433 Changing them to `syscall` instead made it work (not committed here yet), also from a log sample it looks like it _should_ really be singular `"syscall"`:
> 
> ```
> falcosidekick-1    | 2024/01/16 18:51:08 [DEBUG] : Falco's payload : {"uuid":"eaad360f-1c0a-4867-82dc-dde4effc1f5c","output":"18:51:08.842698744: Warning Sensitive file opened for reading by non-trusted p
> rogram (user=root user_loginuid=-1 program=cat command=cat /etc/shadow pid=551252 file=/etc/shadow parent=sh gparent=containerd-shim ggparent=systemd gggparent=\u003cNA\u003e container_id=2cf1d508bf13 ima
> ge=alpine)","priority":"Warning","rule":"Read sensitive file untrusted","time":"2024-01-16T18:51:08.842698744Z","output_fields":{"container.id":"2cf1d508bf13","container.image.repository":"alpine","evt.ti
> me":1705431068842698744,"fd.name":"/etc/shadow","proc.aname[2]":"containerd-shim","proc.aname[3]":"systemd","proc.aname[4]":null,"proc.cmdline":"cat /etc/shadow","proc.name":"cat","proc.pid":551252,"proc.
> pname":"sh","user.loginuid":-1,"user.name":"root"},"source":"syscall","tags":["T1020","T1083","T1212","T1552","T1555","container","filesystem","host","mitre_credential_access","mitre_discovery"],"hostname
> ":"068add2b83db"}   
> ```
> 
> Should we change it to `"syscall"` everywhere in that file (couple more places)?

ping @Issif -^

Issif commented:
> > Did you test my changes? I guess we're closed to the merge, can you rebase on the master and squash your commits after having applied my suggestions. thanks
> 
> @Issif my tests were not working because of your added `"syscalls"` filter in this line [#613 (files)](https://github.com/falcosecurity/falcosidekick/pull/613/files#diff-56307bdb7ded67ebe516f562b90384407b21b2d34e8b5766b42602e1fb9705f0R433) Changing them to `syscall` instead made it work (not committed here yet), also from a log sample it looks like it _should_ really be singular `"syscall"`:
> 
> ```
> falcosidekick-1    | 2024/01/16 18:51:08 [DEBUG] : Falco's payload : {"uuid":"eaad360f-1c0a-4867-82dc-dde4effc1f5c","output":"18:51:08.842698744: Warning Sensitive file opened for reading by non-trusted p
> rogram (user=root user_loginuid=-1 program=cat command=cat /etc/shadow pid=551252 file=/etc/shadow parent=sh gparent=containerd-shim ggparent=systemd gggparent=\u003cNA\u003e container_id=2cf1d508bf13 ima
> ge=alpine)","priority":"Warning","rule":"Read sensitive file untrusted","time":"2024-01-16T18:51:08.842698744Z","output_fields":{"container.id":"2cf1d508bf13","container.image.repository":"alpine","evt.ti
> me":1705431068842698744,"fd.name":"/etc/shadow","proc.aname[2]":"containerd-shim","proc.aname[3]":"systemd","proc.aname[4]":null,"proc.cmdline":"cat /etc/shadow","proc.name":"cat","proc.pid":551252,"proc.
> pname":"sh","user.loginuid":-1,"user.name":"root"},"source":"syscall","tags":["T1020","T1083","T1212","T1552","T1555","container","filesystem","host","mitre_credential_access","mitre_discovery"],"hostname
> ":"068add2b83db"}   
> ```
> 
> Should we change it to `"syscall"` everywhere in that file (couple more places)?

I don't understand why `syscall` works and not `sysdcalls`, see this snippet, this is what falcosidekick set:

```go
	if falcopayload.Source == "" {
		falcopayload.Source = "syscalls"
	}
```

Issif commented on go.mod:
```diff
@@ -1,38 +1,44 @@
 module github.com/falcosecurity/falcosidekick
 
-go 1.20
+go 1.21
```
keep Go 1.20 for now, this upgrade breaks the CI

jjo commented:
> I will create a PR to change that, you will just have to rebase on master then

Thanks, doing

Issif reviewed (dismissed):
:tada:

poiana commented:
LGTM label has been added.  <details>Git tree hash: 309a97a9393afed7a5257a7b4e90675864bbb40b</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falcosidekick/pull/613#pullrequestreview-1865561266" title="Approved">Issif</a>*, *<a href="https://github.com/falcosecurity/falcosidekick/pull/613#" title="Author self-approved">jjo</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalcosidekick).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falcosidekick/blob/master/OWNERS)~~ [Issif]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

jjo commented:
> 🎉

woOOOT ... thanks @Issif especially for the advice and patience towards making this PR a reality 😃!

Issif commented:
> > 🎉
> 
> woOOOT ... thanks @Issif especially for the advice and patience towards making this PR a reality 😃!

it remains 2 lint errors and a build issue, can you check them please

jjo commented:
> > > 🎉
> > 
> > 
> > woOOOT ... thanks @Issif especially for the advice and patience towards making this PR a reality 😃!
> 
> it remains 2 lint errors and a build issue, can you check them please

on them

Issif commented:
/lgtm

poiana commented:
LGTM label has been added.  <details>Git tree hash: f96ce03f839d3a6c1fb80d205b4bf445fa9e24ba</details>

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falcosidekick/pull/613#issuecomment-1930356143" title="LGTM">Issif</a>*, *<a href="https://github.com/falcosecurity/falcosidekick/pull/613#" title="Author self-approved">jjo</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalcosidekick).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falcosidekick/blob/master/OWNERS)~~ [Issif]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->

poiana commented:
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/falcosecurity/falcosidekick/pull/613#pullrequestreview-1865871646" title="Approved">Issif</a>*, *<a href="https://github.com/falcosecurity/falcosidekick/pull/613#" title="Author self-approved">jjo</a>*

The full list of commands accepted by this bot can be found [here](https://prow.falco.org/command-help?repo=falcosecurity%2Ffalcosidekick).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/falcosecurity/falcosidekick/blob/master/OWNERS)~~ [Issif]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>
<!-- META={"approvers":[]} -->
